feature/experimental/p4rt/ate_tests/google_discovery_protocol_packetout_test/packetout_test.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package google_discovery_protocol_packetout_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/cisco-open/go-p4/p4rt_client"
+	"github.com/openconfig/ondatra"
+	p4_v1 "github.com/p4lang/p4runtime/go/p4/v1"
+)
+
+type PacketIO interface {
+	GetPacketOut(t *testing.T, portID uint32, submitIngress bool) []*p4_v1.PacketOut
+}
+
+type testArgs struct {
+	ctx      context.Context
+	leader   *p4rt_client.P4RTClient
+	follower *p4rt_client.P4RTClient
+	dut      *ondatra.DUTDevice
+	ate      *ondatra.ATEDevice
+	top      *ondatra.ATETopology
+	packetIO PacketIO
+}
+
+// sendPackets sends out packets via PacketOut message in StreamChannel.
+func sendPackets(t *testing.T, client *p4rt_client.P4RTClient, packets []*p4_v1.PacketOut, packetCount int) {
+	count := packetCount / len(packets)
+	for _, packet := range packets {
+		for i := 0; i < count; i++ {
+			if err := client.StreamChannelSendMsg(
+				&streamName, &p4_v1.StreamMessageRequest{
+					Update: &p4_v1.StreamMessageRequest_Packet{
+						Packet: packet,
+					},
+				}); err != nil {
+				t.Errorf("There is error seen in Packet Out. %v, %s", err, err)
+			}
+		}
+	}
+}
+
+// testPacketOut sends out PacketOut with GDP payload on p4rt leader or
+// follower client, then verify DUT interface statistics
+func testPacketOut(ctx context.Context, t *testing.T, args *testArgs) {
+	leader := args.leader
+	follower := args.follower
+
+	packetOutTests := []struct {
+		desc       string
+		client     *p4rt_client.P4RTClient
+		expectPass bool
+	}{{
+		desc:       "PacketOut from Primary Controller",
+		client:     leader,
+		expectPass: true,
+	}, {
+		desc:       "PacketOut from Secondary Controller",
+		client:     follower,
+		expectPass: false,
+	}}
+
+	for _, test := range packetOutTests {
+		t.Run(test.desc, func(t *testing.T) {
+			// Check initial packet counters
+			port := sortPorts(args.dut.Ports())[0].Name()
+			counter_0 := args.dut.Telemetry().Interface(port).Counters().OutPkts().Get(t)
+
+			packets := args.packetIO.GetPacketOut(t, portId, false)
+			packet_count := 100
+			sendPackets(t, test.client, packets, packet_count)
+
+			// Wait for ate stats to be populated
+			time.Sleep(60 * time.Second)
+
+			// Check packet counters after packet out
+			counter_1 := args.dut.Telemetry().Interface(port).Counters().OutPkts().Get(t)
```

Comment:
Can we measure the InPkts() in ATE side instead of measuring the DUT out pkts as per the testplan?

Reply:
yes. sure. I will change stats verification from DUT to ATE. thanks.

Reply:
@hariramnat I have done the changes above and also applied similar changes from packetin testcase. Could you please help check if additional changes required?